src/modules/0.0.22/iasql_functions/iasql.ts

```diff
@@ -1037,3 +1039,633 @@ export async function continueUpgrade(
+        context,
+        toCreate,
+        toUpdate,
+        toReplace,
+        toDelete,
+        dryRun,
+        changesByEntity,
+      );
+    }
+  } catch (e: any) {
+    debugObj(e);
+    throw e;
+  } finally {
+    // Create end commit object
+    await insertCommit(orm, dryRun ? 'preview_end' : 'end');
+    orm?.dropConn();
+  }
+}
+
+async function getImportedModules(dbId: string, versionString: string) {
+  const dbMeta = await MetadataRepo.getDbById(dbId);
+  if (dbMeta?.upgrading) throw new Error('Cannot apply a change while upgrading');
+  const importedModules = (AllModules as any)[versionString];
+  if (!importedModules) {
+    throw new Error(`Unsupported version ${versionString}. Please upgrade or replace this database.`);
+  }
+  return importedModules;
+}
+
+async function insertCommit(
+  orm: TypeormWrapper | null,
+  type: 'start' | 'preview_start' | 'end' | 'preview_end',
+) {
+  const commitLog = new IasqlAuditLog();
+  commitLog.user = config.db.user;
+  commitLog.change = {};
+  switch (type) {
+    case 'start':
+      commitLog.changeType = AuditLogChangeType.START_COMMIT;
+      break;
+    case 'preview_start':
+      commitLog.changeType = AuditLogChangeType.PREVIEW_START_COMMIT;
+      break;
+    case 'end':
+      commitLog.changeType = AuditLogChangeType.END_COMMIT;
+      break;
+    case 'preview_end':
+      commitLog.changeType = AuditLogChangeType.PREVIEW_END_COMMIT;
+      break;
+    default:
+      break;
+  }
+  commitLog.tableName = 'iasql_audit_log';
+  commitLog.ts = new Date();
+  await orm?.save(IasqlAuditLog, commitLog);
+  return commitLog;
+}
+
+async function getPreviousStartCommit(orm: TypeormWrapper): Promise<IasqlAuditLog | null> {
+  // Find 'start' commits and pick the second element since the first one should be the one we've just inserted
+  const startCommits = await orm.find(IasqlAuditLog, {
+    order: { ts: 'DESC' },
+    skip: 0,
+    take: 2,
+    where: { changeType: AuditLogChangeType.START_COMMIT },
+  });
+  return startCommits.length > 1 ? startCommits[1] : null;
+}
+
+function getChangesByEntity(changesToCommit: IasqlAuditLog[]): { [key: string]: any[] } {
+  const changesByEntity: { [key: string]: any[] } = {};
+  changesToCommit.forEach((c: IasqlAuditLog) => {
+    const camelCaseEntityName = camelCase(c.tableName);
+    const entityName = camelCaseEntityName.charAt(0).toUpperCase() + camelCaseEntityName.slice(1);
+    const entity: any = {};
+    if (c.changeType === AuditLogChangeType.DELETE) {
+      Object.entries(c.change.original).forEach(([k, v]: [string, any]) => (entity[camelCase(k)] = v));
+    } else if ([AuditLogChangeType.INSERT, AuditLogChangeType.UPDATE].includes(c.changeType)) {
+      Object.entries(c.change.change).forEach(([k, v]: [string, any]) => (entity[camelCase(k)] = v));
+    }
+    if (changesByEntity[entityName]) {
+      changesByEntity[entityName].push(entity);
+    } else {
+      changesByEntity[entityName] = [entity];
+    }
```

Comment:
Nit: This could just be

```suggestion
    changesByEntity[entityName] = changesByEntity[entityName] ?? [];
    changesByEntity[entityName].push(entity);
```